### ISSUE TITLE
velocity limit

### ISSUE THREAD

**Issue by jordangarside** [IjzerenHein/famous-flex#105]
Is there any way to add or impose a velocity limit presently?
If not, would it be possible to add?

I'm doing infinite scrolling and if I allow any velocity then the user can reach the end of the page before the next set of results has loaded.

**Comment by IjzerenHein** [comment]
Hi, yes that could be added. However, it does seem like a sort of hackish solution to the problem you are having. Using timeouts and in this case a max-velocity to ensure that end of the content isn't reached, feels like something that will bite you in the ass at the end. Also it will decrement the user-experience in imho because the actual velocity will be different than the finger-release velocity. And what if the server is a bit busy and scrollable content is delivered at a slightly later moment?!

Anyway, you can easily implement it yourself, like this:

```javascript
var scrollView = new FlexScrollView({..});
scrollView.on('swipeend', function() {
   var velocity = scrollView.getVelocity();
   velocity = Math.max(Math.min(velocity, 0.5), -0.5);
   scrollView.setVelocity(velocity);
});
```

**Comment by jordangarside** [comment]
It's being done in Meteor with Meteor-Famous-Views, by observing the bottom element and requesting more results from the server. The problem is that, at present, famous-views can't render the content/surfaces quickly enough if you scroll as fast as possible, this then makes it so the user hits the "end" of the content because the new surfaces are still rendering. This takes away the experience of infinite scrolling.
I have a buffer of about 60 elements, and request when the user is still 50 elements away from the bottom (using collection-layout).

**Comment by Wenape** [comment]
It would be nice to have a `maxVelocity` option when initializing the FlexScrollview. When the scroll speed becomes too fast (usually after a really, really fast swipe), even if the browser is not doing much scripting or rendering, painting alone can cause the app to become choppy (depending on the content in the view).

I realize that reaching higher velocities that would cause this is not the usual case, but it's quite common, for example, if you use a method like `ensureVisible`. If the element is way down the scrollview, then the scrolling speed used to get there is way too much and causes choppiness.

**Comment by IjzerenHein** [comment]
Hi, I've just added a way to impose a velocity cap to the 'develop' branch:
https://github.com/IjzerenHein/famous-flex/commit/f7bd51401e8a42a36adb9f48165e24ded3568a0f

The velocity can be capped in the physics engine of famo.us.  To configure it use:

```javascript
var scrollView = new FlexScrollView({
  scrollPhysicsEngine: {
    velocityCap: number
  }
});
```

I don't know what kind of number would be a good one to use, you'll have to experiment with that.

cheers

**Comment by Wenape** [comment]
Thanks @IjzerenHein for taking a look into this. Just tested the `velocityCap` option you added, it works great. I would say an optimal velocity would have to be on a range of `[0 - 5]` to get a good performance around all devices.

**Comment by IjzerenHein** [comment]
Awesome! Let me know which velocity-cap you choose, I'll consider putting it in as a default.

**Comment by IjzerenHein** [comment]
ping @Wenape

**Comment by Wenape** [comment]
Sorry for the delay. I've been thinking about a good default value that could be used. I'm not sure the one that I use would be the best one as it is not the usual use case.

There's a bunch of things that can affect the scrolling performance independent from the device. The two extreme cases would be: (Fastest: Minimal Rendering + Minimal Painting) A scrollview with elements that just have text, maybe a small image and uses `layoutAll=true` option, so everything renders at the beginning. (Slowest: Rendering Involved + A lot of Painting) A scrollview which uses full image containers and uses the default rendering functionality (New elements render as they come into view). 

Between this two an optimal velocity cap would greatly differ. I would say you could get away with not even using a velocity-cap for the first case & for the second having to use something as low as 3-4 velocity to get a good performance.

I've seen the case for most people is so use a "contacts like" scrollview. This means elements with some text and a small avatar image. Using the default option `layoutAll=false` as usually too many elements are involved. For this case, a velocity-cap of  `10` would be the one that offers the best performance with out being restrictive. It also works well with the different scrollTo methods of the flexScrollView. 

With other specific cases it's up to the user to use a different value depending on the scrollview's content.

**Comment by IjzerenHein** [comment]
Yeah I agree with you completely, choosing a right limit might depend on the content and speed of the device/network. That's a tough cookie to crack to be honest. Ideally we would need something that creates a sort of "natural" throttling effect, but I wouldn't know how to implement that.

Anyway, thanks very much for the detailed explanation!

The option to configure the velocity-cap will be released in the next v0.3.x version. After that it will probably be mixed-mode that I will be working on.

**Comment by IjzerenHein** [comment]
Hi, for your information, the ability to configure the velocityCap has been released in an official version:
https://github.com/IjzerenHein/famous-flex/releases/tag/v0.3.5
cheers
